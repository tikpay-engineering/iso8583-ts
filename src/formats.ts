export type NumericEncoding = 'bcd' | 'ascii'
export type LenHeaderEncoding = 'bcd' | 'ascii'
