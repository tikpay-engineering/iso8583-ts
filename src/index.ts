export { Iso8583 } from './iso8583'
export type { FieldSpec, MessageSpec } from './iso8583'

export {
  A,
  AN,
  ANS,
  B,
  bitmap,
  Kind,
  LLLVAR,
  LLLVARan,
  LLLVARans,
  LLLVARn,
  LLVAR,
  LLVARan,
  LLVARans,
  LLVARn,
  N,
  NumericEncoding,
  VarLenCountMode,
  VarLenHeaderEncoding,
  VarPayloadEncoding,
} from './internals/formats'
