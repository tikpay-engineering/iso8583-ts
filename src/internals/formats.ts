export type NumericEncoding = 'bcd' | 'ascii'
export type LenHeaderEncoding = 'bcd' | 'ascii'
export type VarEncoding = 'ascii' | 'binary' | 'bcd-digits'
export type VarLenCount = 'bytes' | 'digits'

export enum Kind {
  Alpha = 'a',
  AlphaNumeric = 'an',
  AlphaNumericSpecial = 'ans',
  Binary = 'b',
  LLVAR = 'LLVAR',
  LLVARn = 'LLVARn',
  LLVARan = 'LLVARan',
  LLVARans = 'LLVARans',
  LLLVAR = 'LLLVAR',
  LLLVARn = 'LLLVARn',
  LLLVARan = 'LLLVARan',
  LLLVARans = 'LLLVARans',
  Numeric = 'n',
}

export type AFormat = { kind: Kind.Alpha; length: number }
export type ANFormat = { kind: Kind.AlphaNumeric; length: number }
export type ANSFormat = { kind: Kind.AlphaNumericSpecial; length: number }
export type BFormat = { kind: Kind.Binary; length: number }
export type NFormat = { kind: Kind.Numeric; length: number; encoding?: NumericEncoding }

type VarBase = {
  length: number
  payload?: VarEncoding
  lenHeader?: LenHeaderEncoding
  lenCounts?: VarLenCount
}

export type LLVARFormat = { kind: Kind.LLVAR | Kind.LLVARn | Kind.LLVARan | Kind.LLVARans } & VarBase

export type LLLVARFormat = { kind: Kind.LLLVAR | Kind.LLLVARn | Kind.LLLVARan | Kind.LLLVARans } & VarBase

export type VARFormatRequired = (LLVARFormat | LLLVARFormat) &
  Required<Pick<VarBase, 'payload' | 'lenHeader' | 'lenCounts'>>
