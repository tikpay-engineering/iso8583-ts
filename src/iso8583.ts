import { buildBitmap, parseBitmap } from '@bits/bitmap'
import { assertMTI, decodeMTI, encodeMTI } from '@bits/mti'
import { decodeAlpha, encodeAlpha } from '@encodings/alpha'
import { decodeBinary, encodeBinary } from '@encodings/binary'
import { decodeNumeric, encodeNumeric } from '@encodings/numeric'
import { decodeVar, encodeVar } from '@encodings/varlen'
import { BitmapConstraint, ERR } from '@internals/constants'
import { FormatObject, Kind, NumericEncoding } from '@internals/formats'

export type EncoderOptions = {
  mtiEncoding?: NumericEncoding
}

export type FieldSpec = { name: string; format: FormatObject }
export type MessageSpec = Record<number, FieldSpec>

export type PackedMessage = { mti: string; bytes: Buffer }
export type UnpackedMessage = { mti: string; fields: Record<number, Buffer | string | number>; bytesRead: number }

export class Iso8583 {
  private spec: MessageSpec
  private mtiEncoding: NumericEncoding = NumericEncoding.BCD
  private bitmapConstraint: BitmapConstraint = 64

  constructor(spec: MessageSpec) {
    this.spec = spec

    const mtiField = this.spec[0]
    if (mtiField) {
      if (mtiField.format.kind !== Kind.Numeric) throw new Error(ERR.INVALID_MTI_SPEC(0))
      this.mtiEncoding = mtiField.format.encoding || NumericEncoding.BCD
    }

    const bmField = this.spec[1]
    if (bmField) {
      if (bmField.format.kind !== Kind.Bitmap) throw new Error(ERR.INVALID_BITMAP_SPEC(1))
      this.bitmapConstraint = bmField.format.length === 8 ? 64 : 128
    }
  }

  pack = (mti: string, fields: Record<number, Buffer | string | number>): PackedMessage => {
    assertMTI(mti)
    const present = Object.keys(fields)
      .map(Number)
      .filter(n => n >= 2 && n <= 128)
      .sort((a, b) => a - b)
    for (const de of present) if (!this.spec[de]) throw new Error(ERR.NO_SPEC(de))

    const chunks: Buffer[] = []
    chunks.push(encodeMTI(mti, this.mtiEncoding))
    chunks.push(buildBitmap(present, this.bitmapConstraint))

    for (const de of present) {
      chunks.push(this.encodeField(de, this.spec[de], fields[de]))
    }

    return { mti, bytes: Buffer.concat(chunks) }
  }

  unpack = (buf: Buffer): UnpackedMessage => {
    let offset = 0
    const { mti, read } = decodeMTI(buf, offset, this.mtiEncoding)
    offset += read
    assertMTI(mti)

    if (buf.length - offset < 8) throw new Error(ERR.PRIMARY_UNDERRUN)
    const primary = buf.subarray(offset, offset + 8)
    offset += 8

    const hasSecondary = (primary[0] & 0x80) !== 0
    let bitmap = primary
    if (hasSecondary) {
      if (this.bitmapConstraint === 64) throw new Error(ERR.SEC_BITMAP_CONSTRAINED)
      if (buf.length - offset < 8) throw new Error(ERR.SEC_UNDERRUN)
      bitmap = Buffer.concat([primary, buf.subarray(offset, offset + 8)])
      offset += 8
    }

    const present = parseBitmap(bitmap, this.bitmapConstraint)
    const fields: Record<number, Buffer | string | number> = {}
    for (const de of present) {
      const spec = this.spec[de]
      if (!spec) throw new Error(ERR.NO_SPEC(de))
      const { value, read } = this.decodeField(de, spec, buf, offset)
      fields[de] = value
      offset += read
    }

    return { mti, fields, bytesRead: offset }
  }

  explain = (buf: Buffer): string => {
    const decoded = this.unpack(buf)
    const lines = [`MTI: ${decoded.mti}`]
    for (const [idStr, value] of Object.entries(decoded.fields)) {
      const id = Number(idStr)
      const spec = this.spec[id]
      const { kind, length } = spec.format
      lines.push(`${id.toString().padStart(3, '0')} ${spec.name} (${kind}, len=${length}): ${value}`)
    }
    return lines.join('\n')
  }

  encodeField(de: number, spec: FieldSpec, value: Buffer | string | number): Buffer {
    const f = spec.format
    switch (f.kind) {
      case Kind.Numeric:
        return encodeNumeric(f, value as string | number)
      case Kind.Alpha:
      case Kind.AlphaNumeric:
      case Kind.AlphaNumericSpecial:
        return encodeAlpha(de, f, value as string)
      case Kind.Binary:
        return encodeBinary(de, f, value as Buffer | string)
      case Kind.LLVAR:
      case Kind.LLVARn:
      case Kind.LLVARan:
      case Kind.LLVARans:
      case Kind.LLLVAR:
      case Kind.LLLVARn:
      case Kind.LLLVARan:
      case Kind.LLLVARans:
        return encodeVar(de, f, value as Buffer | string)
      default:
        throw new Error(ERR.UNSUPPORTED(f.kind))
    }
  }

  decodeField(de: number, spec: FieldSpec, buf: Buffer, offset: number) {
    const f = spec.format
    switch (f.kind) {
      case Kind.Numeric:
        return decodeNumeric(f, buf, offset)
      case Kind.Alpha:
      case Kind.AlphaNumeric:
      case Kind.AlphaNumericSpecial:
        return decodeAlpha(f, buf, offset)
      case Kind.Binary:
        return decodeBinary(de, f, buf, offset)
      case Kind.LLVAR:
      case Kind.LLVARn:
      case Kind.LLVARan:
      case Kind.LLVARans:
      case Kind.LLLVAR:
      case Kind.LLLVARn:
      case Kind.LLLVARan:
      case Kind.LLLVARans:
        return decodeVar(de, f, buf, offset)
      default:
        throw new Error(ERR.UNSUPPORTED(f.kind))
    }
  }
}
