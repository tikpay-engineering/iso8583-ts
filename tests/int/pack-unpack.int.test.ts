import { Kind, NumericEncoding } from '@internals/formats'
import { Iso8583, MessageSpec } from '../../src/iso8583'

const messageSpec: MessageSpec = {
  0: { name: 'MTI', format: { kind: Kind.Numeric, length: 4, encoding: NumericEncoding.BCD } },
  1: { name: 'Bitmap', format: { kind: Kind.Bitmap, length: 8 } },

  2: { name: 'PAN (LLVARn)', format: { kind: Kind.LLVARn, length: 19 } },
  3: { name: 'Processing Code', format: { kind: Kind.Numeric, length: 6, encoding: NumericEncoding.BCD } },
  4: { name: 'Amount', format: { kind: Kind.Numeric, length: 12, encoding: NumericEncoding.BCD } },
  5: { name: 'Alpha field', format: { kind: Kind.Alpha, length: 5 } },
  6: { name: 'Alphanumeric', format: { kind: Kind.AlphaNumeric, length: 10 } },
  7: { name: 'AlphanumericSpec', format: { kind: Kind.AlphaNumericSpecial, length: 12 } },
  8: { name: 'Binary data', format: { kind: Kind.Binary, length: 4 } },

  20: { name: 'LLVAR alpha', format: { kind: Kind.LLVARan, length: 20 } },
  21: { name: 'LLVAR ans', format: { kind: Kind.LLVARans, length: 20 } },
  22: { name: 'LLLVAR alpha', format: { kind: Kind.LLLVARan, length: 999 } },
  23: { name: 'LLLVAR ans', format: { kind: Kind.LLLVARans, length: 999 } },
  24: { name: 'LLLVAR numeric', format: { kind: Kind.LLLVARn, length: 999 } },
}

const normalize = (v: string | number | Buffer) => (Buffer.isBuffer(v) ? v.toString('hex') : String(v))

describe('integration: pack + unpack with all formats', () => {
  const iso = new Iso8583(messageSpec)

  const fields = {
    2: '4761731234567890',
    3: '000000',
    4: '000000012345',
    5: 'ABCDE',
    6: 'ABC123XYZ9',
    7: 'ABCDE*FGHIJ!',
    8: Buffer.from('abcd', 'utf8'),
    20: 'ALPHA_VAR',
    21: 'ANS_VAR!@#',
    22: 'ALPHA_LONG_VAR',
    23: 'ANS_LONG_VAR*&^',
    24: '1234567890',
  }

  it('round-trips a message with one of every format', () => {
    const packed = iso.pack('0200', fields)
    const unpacked = iso.unpack(packed.bytes)

    expect(unpacked.mti).toBe('0200')

    const actual = Object.fromEntries(Object.entries(unpacked.fields).map(([k, v]) => [k, normalize(v)]))
    const expected = Object.fromEntries(Object.entries(fields).map(([k, v]) => [k, normalize(v)]))

    expect(actual).toEqual(expected)
  })

  it('produces a readable explain output with snapshot check', () => {
    const packed = iso.pack('0200', fields)
    const explain = iso.explain(packed.bytes)

    expect(explain).toContain('MTI: 0200')
    expect(explain).toContain('002 PAN (LLVARn)')
    expect(explain).toContain('003 Processing Code')
    expect(explain).toContain('005 Alpha field')
    expect(explain).toContain('008 Binary data')

    expect(explain).toContain('ABCDE')
    expect(explain).toContain('ABC123XYZ9')
    expect(explain).toContain('ALPHA_LONG_VAR')
    expect(explain).toContain('1234567890')

    expect(explain).toMatchSnapshot()
  })
})
