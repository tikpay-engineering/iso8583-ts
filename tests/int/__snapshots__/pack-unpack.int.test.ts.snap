// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration: pack + unpack with all formats > produces a readable explain output with snapshot check 1`] = `
"MTI: 0200
002 PAN (LLVARn) (LLVARn, len=19): 4761731234567890
003 Processing Code (n, len=6): 000000
004 Amount (n, len=12): 000000012345
005 Alpha field (a, len=5): ABCDE
006 Alphanumeric (an, len=10): ABC123XYZ9
007 AlphanumericSpec (ans, len=12): ABCDE*FGHIJ!
008 Binary data (b, len=4): abcd
020 LLVAR alpha (LLVARan, len=20): ALPHA_VAR
021 LLVAR ans (LLVARans, len=20): ANS_VAR!@#
022 LLLVAR alpha (LLLVARan, len=999): ALPHA_LONG_VAR
023 LLLVAR ans (LLLVARans, len=999): ANS_LONG_VAR*&^
024 LLLVAR numeric (LLLVARn, len=999): 1234567890"
`;
